test/jdk/java/net/httpclient/SendResponseHeadersTest.java

```diff
@@ -0,0 +1,102 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8253005
+ * @summary Check that sendResponseHeaders throws an IOException when headers
+ *  have already been sent
+ * @run testng/othervm SendResponseHeadersTest
+ */
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.testng.Assert.expectThrows;
+import static org.testng.Assert.fail;
+
+
+public class SendResponseHeadersTest {
+
+    @Test
+    public void testSend() throws Exception {
+        var addr = new InetSocketAddress(InetAddress.getLoopbackAddress(), 0);
+        var server = HttpServer.create(addr, 0);
+        var path = "/test/foo.html";
+        server.createContext("/test", new TestHandler());
+        ExecutorService executor = Executors.newCachedThreadPool();
+        server.setExecutor (executor);
+        server.start();
+
+        URI uri = new URI("http", null,
+                server.getAddress().getHostString(),
+                server.getAddress().getPort(),
+                path, null, null);
+
+        try {
+            HttpClient client = HttpClient.newHttpClient();
+            HttpRequest request = HttpRequest.newBuilder(uri)
+                    .GET()
+                    .build();
+            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+            // verify empty response received, otherwise an error has occurred
+            if (!response.body().isEmpty())
+                fail(response.body());
+        } finally {
+            server.stop(2);
+            executor.shutdown();
+        }
+    }
+
+    static class TestHandler implements HttpHandler {
+        public void handle(HttpExchange exchange) throws IOException {
+            exchange.sendResponseHeaders(200, 0);
```

Comment:
It's always preferable to read the request body fully before calling `sendResponseHeaders`

Reply:
Thanks for pointing this out. I've added it to the test, and you can view it in commit https://github.com/openjdk/jdk/pull/1014/commits/814a7fd014d7706f63ed239f8be56ff26190b1f5